File: java/lib/src/main/java/dev/argon/esexpr/ESExprBinaryReader.java
package dev.argon.esexpr;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.stream.Stream;

/**
 * A reader for the ESExpr binary format.
 */
public class ESExprBinaryReader {
	/**
	 * Create a reader for the ESExpr binary format.
	 * @param symbolTable The symbol table used when parsing.
	 * @param is The stream.
	 */
	public ESExprBinaryReader(@NotNull List<String> symbolTable, @NotNull InputStream is) {
		this.symbolTable = new ArrayList<>(symbolTable);
		this.is = is;
	}

	private final List<String> symbolTable;
	private final @NotNull InputStream is;
	private int nextByte = -1;

	/**
	 * Attempts to read an ESExpr from the stream.
	 * @return The ESExpr, or null if at the end of the stream.
	 * @throws IOException when an error occurs in the underlying stream.
	 * @throws SyntaxException when an expression cannot be read.
	 */
	public @Nullable ESExpr read() throws IOException, SyntaxException {
		if(peekNext() < 0) {
			return null;
		}

		return readExpr();
	}

	/**
	 * Reads all ESExpr values from the stream.
	 * @return A stream of ESExpr values.
	 */
	public @NotNull Stream<@NotNull ESExpr> readAll() {
		return Stream
			.generate(() -> {
				try {
					return read();
				}
				catch(IOException | SyntaxException ex) {
					throw new RuntimeException(ex);
				}
			})
			.takeWhile(Objects::nonNull);
	}

	/**
	 * Reads all ESExpr values, using the first as the string table.
	 * @param is The input stream.
	 * @return A stream of ESExpr values.
	 * @throws IOException when an error occurs in the underlying stream.
	 * @throws SyntaxException when an expression cannot be read.
	 */
	public static @NotNull Stream<@NotNull ESExpr> readEmbeddedStringTable(InputStream is) throws IOException, SyntaxException {
		try {
			var stExpr = new ESExprBinaryReader(List.of(), is).readExpr();
			var stringTable = StringTable.codec.decode(stExpr);
			return new ESExprBinaryReader(stringTable.values(), is).readAll();
		}
		catch(DecodeException ex) {
			throw new SyntaxException(ex);
		}
	}



	private int next() throws IOException {
		if(nextByte >= 0) {
			int res = nextByte;
			nextByte = -1;
			return res;
		}

		return is.read();
	}

	private int peekNext() throws IOException {
		if(nextByte >= 0) {
			return nextByte;
		}

		nextByte = is.read();
		return nextByte;
	}

	private BinToken nextToken() throws IOException, SyntaxException {
		int b = next();
		if(b < 0) {
			throw new EOFException();
		}

		BinToken.WithIntegerType type = switch((b & 0xE0)) {
			case 0x00 -> BinToken.WithIntegerType.CONSTRUCTOR;
			case 0x20 -> BinToken.WithIntegerType.INT;
			case 0x40 -> BinToken.WithIntegerType.NEG_INT;
			case 0x60 -> BinToken.WithIntegerType.STRING;
			case 0x80 -> BinToken.WithIntegerType.STRING_POOL_INDEX;
			case 0xA0 -> BinToken.WithIntegerType.BINARY;
			case 0xC0 -> BinToken.WithIntegerType.KEYWORD;
			default -> null;
		};

		if(type == null) {
			return switch(b) {
				case 0xE0 -> BinToken.Fixed.CONSTRUCTOR_END;
				case 0xE1 -> BinToken.Fixed.TRUE;
				case 0xE2 -> BinToken.Fixed.FALSE;
				case 0xE3 -> BinToken.Fixed.NULL;
				case 0xE4 -> BinToken.Fixed.FLOAT32;
				case 0xE5 -> BinToken.Fixed.FLOAT64;
				case 0xE6 -> BinToken.Fixed.CONSTRUCTOR_START_STRING_TABLE;
				case 0xE7 -> BinToken.Fixed.CONSTRUCTOR_START_LIST;
				default -> throw new SyntaxException();
			};
		}
		else {
			BigInteger i = BigInteger.valueOf(b & 0x0F);
			if((b & 0x10) == 0x10) {
				i = readInt(i, 4);
			}

			return new BinToken.WithInteger(type, i);
		}
	}

	private BigInteger readInt(BigInteger acc, int bits) throws IOException {
		while(true) {
			int b = next();
			if(b < 0) {
				throw new EOFException();
			}

			acc = acc.or(BigInteger.valueOf(b & 0x7F).shiftLeft(bits));
			bits += 7;

			if((b & 0x80) == 0) {
				return acc;
			}
		}
	}



	private @NotNull ESExpr readExpr() throws SyntaxException, IOException {
		return switch(readExprPlus()) {
			case ExprPlus.Expr(var expr) -> expr;
			default -> throw new SyntaxException();
		};
	}

	private sealed interface ExprPlus {
		record Expr(ESExpr expr) implements ExprPlus {}
		record ConstructorEnd() implements ExprPlus {}
		record Keyword(String name) implements ExprPlus {}
	}

	private @NotNull ExprPlus readExprPlus() throws SyntaxException, IOException {
		return switch(nextToken()) {
			case BinToken.WithInteger(var type, var value) -> switch(type) {
				case CONSTRUCTOR -> {
					var sym = symbolTable.get(value.intValueExact());
					yield new ExprPlus.Expr(readConstructor(sym));
				}
				case INT -> new ExprPlus.Expr(new ESExpr.Int(value));
				case NEG_INT -> new ExprPlus.Expr(new ESExpr.Int(value.add(BigInteger.ONE).negate()));

				// Should be safe to bypass next/peekNext here.
				case STRING -> {
					int len = value.intValueExact();
					byte[] b = new byte[len];
					if(is.readNBytes(b, 0, len) < len) {
						throw new EOFException();
					}

					yield new ExprPlus.Expr(new ESExpr.Str(new String(b, StandardCharsets.UTF_8)));
				}

				case STRING_POOL_INDEX -> {
					var sym = symbolTable.get(value.intValueExact());
					yield new ExprPlus.Expr(new ESExpr.Str(sym));
				}

				case BINARY -> {
					int len = value.intValueExact();
					byte[] b = new byte[len];
					if(is.readNBytes(b, 0, len) < len) {
						throw new EOFException();
					}

					yield new ExprPlus.Expr(new ESExpr.Binary(b));
				}

				case KEYWORD -> {
					var sym = symbolTable.get(value.intValueExact());
					yield new ExprPlus.Keyword(sym);
				}
			};

			case BinToken.Fixed fixed -> switch(fixed) {
				case NULL -> new ExprPlus.Expr(new ESExpr.Null());
				case CONSTRUCTOR_END -> new ExprPlus.ConstructorEnd();
				case TRUE -> new ExprPlus.Expr(new ESExpr.Bool(true));
				case FALSE -> new ExprPlus.Expr(new ESExpr.Bool(false));
				case FLOAT32 -> {
					int bits = 0;
					for(int i = 0; i < 4; ++i) {
						int b = next();
						if(b < 0) {
							throw new EOFException();
						}

						bits |= (b & 0xFF) << (i * 8);
					}

					yield new ExprPlus.Expr(new ESExpr.Float32(Float.intBitsToFloat(bits)));
				}
				case FLOAT64 -> {
					long bits = 0;
					for(int i = 0; i < 8; ++i) {
						int b = next();
						if(b < 0) {
							throw new EOFException();
						}

						bits |= (long)(b & 0xFF) << (i * 8);
					}

					yield new ExprPlus.Expr(new ESExpr.Float64(Double.longBitsToDouble(bits)));
				}

				case CONSTRUCTOR_START_STRING_TABLE -> new ExprPlus.Expr(readConstructor(BinToken.StringTableName));
				case CONSTRUCTOR_START_LIST -> new ExprPlus.Expr(readConstructor(BinToken.ListName));
			};
		};
	}

	private @NotNull ESExpr readConstructor(String name) throws IOException, SyntaxException {
		var args = new ArrayList<ESExpr>();
		var kwargs = new HashMap<String, ESExpr>();

		body:
		while(true) {
			switch(readExprPlus()) {
				case ExprPlus.Expr(var expr) -> args.add(expr);
				case ExprPlus.ConstructorEnd() -> {
					break body;
				}
				case ExprPlus.Keyword(var kw) -> {
					var expr = readExpr();
					kwargs.put(kw, expr);
				}
			}
		}

		return new ESExpr.Constructor(name, args, kwargs);
	}

}

File: dotnet/ESExpr.Runtime/ESExprBinaryReader.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.Linq;
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using ESExpr.Runtime.Codecs;

namespace ESExpr.Runtime;

public class ESExprBinaryReader {
	public ESExprBinaryReader(Stream stream) {
		symbolTable = new List<string>();
		this.stream = stream;
	}

	public ESExprBinaryReader(IImmutableList<string> symbolTable, Stream stream) {
		this.symbolTable = symbolTable.ToList();
		this.stream = stream;
	}

	private readonly List<string> symbolTable;
	private readonly Stream stream;

	private readonly byte[] byteBuffer = new byte[1];
	private int nextByte = -1;

	public async ValueTask<Expr?> TryRead(CancellationToken cancellationToken = default) {
		var readVisitor = new ReadExprVisitor();

		for(; ; )
		{
			cancellationToken.ThrowIfCancellationRequested();

			var expr = await ReadExprPlus(readVisitor, cancellationToken);
			if(readVisitor.IsDone) {
				return expr;
			}
		}
	}

	public async ValueTask<Expr> Read(CancellationToken cancellationToken = default) {
		var expr = await TryRead(cancellationToken).ConfigureAwait(false);
		if(expr is null) {
			throw new EndOfStreamException();
		}

		return expr;
	}

	public async IAsyncEnumerable<Expr> ReadAll([EnumeratorCancellation] CancellationToken cancellationToken = default) {
		while(true) {
			cancellationToken.ThrowIfCancellationRequested();

			var expr = await TryRead(cancellationToken).ConfigureAwait(false);
			if(expr is null) {
				break;
			}

			yield return expr;
		}
	}

	private string LookupStringTable(BigInteger index) {
		if(index < 0 || index >= symbolTable.Count) {
			throw new SyntaxException();
		}

		return symbolTable[(int)index];
	}


	private async ValueTask<int> Next(CancellationToken cancellationToken = default) {
		if(nextByte >= 0) {
			int res = nextByte;
			nextByte = -1;
			return res;
		}

		int bytesRead = await stream.ReadAsync(byteBuffer, cancellationToken).ConfigureAwait(false);
		if(bytesRead == 0) {
			return -1;
		}

		return byteBuffer[0];
	}

	private interface IExprPlusVisitor<out T> {
		T VisitExpr(Expr expr);
		T VisitConstructorEnd();
		T VisitKeyword(BigInteger index);
		T VisitAppendedToStringTable();
		T VisitEndOfStream();
	}

	private sealed class ReadExprVisitor : IExprPlusVisitor<Expr?> {
		public ReadExprVisitor() { }

		public bool IsDone { get; private set; } = false;

		public Expr? VisitExpr(Expr expr) {
			IsDone = true;
			return expr;
		}

		public Expr? VisitConstructorEnd() {
			throw new SyntaxException();
		}

		public Expr? VisitKeyword(BigInteger index) {
			throw new SyntaxException();
		}

		public Expr? VisitAppendedToStringTable() {
			return null;
		}

		public Expr? VisitEndOfStream() {
			IsDone = true;
			return null;
		}
	}

	private async ValueTask<T> ReadExprPlus<T>(IExprPlusVisitor<T> visitor, CancellationToken cancellationToken) {
		var tokenOpt = await NextToken(cancellationToken).ConfigureAwait(false);
		if(tokenOpt is not { } token) {
			return visitor.VisitEndOfStream();
		}

		Expr expr;
		switch(token.BinTokenType) {
			case BinToken.TokenType.Constructor:
				expr = await ReadConstructor(LookupStringTable(token.IntValue ?? throw new SyntaxException()), cancellationToken).ConfigureAwait(false);
				break;

			case BinToken.TokenType.Int:
				expr = new Expr.Int(token.IntValue ?? throw new SyntaxException());
				break;

			case BinToken.TokenType.NegInt:
				expr = new Expr.Int(-((token.IntValue ?? throw new SyntaxException()) + 1));
				break;

			case BinToken.TokenType.String: {
				var bytes = await ReadBytes(token.IntValue ?? throw new SyntaxException(), cancellationToken);
				expr = new Expr.Str(Encoding.UTF8.GetString(bytes));
				break;
			}

			case BinToken.TokenType.StringPoolIndex:
				expr = new Expr.Str(LookupStringTable(token.IntValue ?? throw new SyntaxException()));
				break;

			case BinToken.TokenType.Null0:
				expr = new Expr.Null(0);
				break;

			case BinToken.TokenType.Null1:
				expr = new Expr.Null(1);
				break;

			case BinToken.TokenType.Null2:
				expr = new Expr.Null(2);
				break;

			case BinToken.TokenType.NullN: {
				var level = await ReadInt(0, 0, cancellationToken).ConfigureAwait(false);
				expr = new Expr.Null(level + 3);
				break;
			}


			case BinToken.TokenType.True:
				expr = new Expr.Bool(true);
				break;

			case BinToken.TokenType.False:
				expr = new Expr.Bool(false);
				break;

			case BinToken.TokenType.Float16: {
				var bits = await ReadFixed<ushort>(sizeof(ushort), cancellationToken);
				var value = BitConverter.UInt16BitsToHalf(bits);
				if(Half.IsNaN(value)) {
					expr = new Expr.Float16NaN(bits);					
				}
				else {
					expr = new Expr.Float16(value);
				}
				
				break;
			}

			case BinToken.TokenType.Float32: {
				var bits = await ReadFixed<uint>(sizeof(float), cancellationToken);
				var value = BitConverter.UInt32BitsToSingle(bits);

				if(float.IsNaN(value)) {
					expr = new Expr.Float32NaN(bits);
				}
				else {
					expr = new Expr.Float32(value);	
				}
				break;
			}

			case BinToken.TokenType.Float64: {
				var bits = await ReadFixed<ulong>(sizeof(double), cancellationToken);
				var value = BitConverter.UInt64BitsToDouble(bits);

				if(double.IsNaN(value)) {
					expr = new Expr.Float64NaN(bits);
				}
				else {
					expr = new Expr.Float64(value);
				}
				break;
			}

			case BinToken.TokenType.Array8:
				expr = new Expr.Array8(
					(await ReadBytes(token.IntValue ?? throw new SyntaxException(), cancellationToken)).ToImmutableArray()
				);
				break;

			case BinToken.TokenType.Array16: {
				expr = new Expr.Array16(await ReadArrayN<ushort>(sizeof(ushort), cancellationToken));
				break;
			}

			case BinToken.TokenType.Array32: {
				expr = new Expr.Array32(await ReadArrayN<uint>(sizeof(uint), cancellationToken));
				break;
			}

			case BinToken.TokenType.Array64: {
				expr = new Expr.Array64(await ReadArrayN<ulong>(sizeof(ulong), cancellationToken));
				break;
			}

			case BinToken.TokenType.Array128: {
				expr = new Expr.Array128(await ReadArrayN<UInt128>(16, cancellationToken));
				break;
			}


			case BinToken.TokenType.ConstructorStartStringTable:
				expr = await ReadConstructor(StringTable.StringTableConstructor, cancellationToken).ConfigureAwait(false);
				break;

			case BinToken.TokenType.ConstructorStartList:
				expr = await ReadConstructor(ListCodecBase<int, List<int>>.ListConstructor, cancellationToken).ConfigureAwait(false);
				break;
			
			case BinToken.TokenType.ConstructorStartMap:
				expr = await ReadConstructor(MapCodecBase<int, int, Dictionary<int, int>>.MapConstructor, cancellationToken).ConfigureAwait(false);
				break;
			
			case BinToken.TokenType.ConstructorStartSet:
				expr = await ReadConstructor(SetCodecBase<int, HashSet<int>>.SetConstructor, cancellationToken).ConfigureAwait(false);
				break;

			case BinToken.TokenType.AppendStringTable: {
				var newStringTableExpr = await Read(cancellationToken).ConfigureAwait(false);

				if(newStringTableExpr is Expr.Str s) {
					symbolTable.Add(s.value);
				}
				else {
					StringTable newStringTable;
					try {
						IESExprCodec<StringTable> codec = StringTable.Codec;
						newStringTable = codec.Decode(newStringTableExpr);
					}
					catch(DecodeException ex) {
						throw new SyntaxException("Could not decode string table", ex);
					}

					symbolTable.AddRange(newStringTable.strings);
				}

				return visitor.VisitAppendedToStringTable();
			}

			case BinToken.TokenType.ConstructorEnd:
				return visitor.VisitConstructorEnd();

			case BinToken.TokenType.Keyword:
				return visitor.VisitKeyword(token.IntValue ?? throw new SyntaxException());

			default:
				throw new SyntaxException();
		}

		return visitor.VisitExpr(expr);
	}

	private async ValueTask<N> ReadFixed<N>(int size, CancellationToken cancellationToken = default)
		where N : IBinaryInteger<N> {
		var bytes = await ReadBytes(size, cancellationToken);
		N value = N.Zero;
		for(int i = 0; i < size; ++i) {
			value |= N.CreateTruncating(bytes[i]) << (i * 8);
		}

		return value;
	}

	private async ValueTask<ImmutableArray<T>> ReadArrayN<T>(int byteSize, CancellationToken cancellationToken = default)
		where T : IBinaryInteger<T> {
		var lengthBig = await ReadInt(0, 0, cancellationToken).ConfigureAwait(false);
		var buff = await ReadBytes(lengthBig * byteSize, cancellationToken);
		int length = (int)lengthBig;
		var builder = ImmutableArray.CreateBuilder<T>(length);
		for(int i = 0; i < length; ++i) {
			T value = T.Zero;
			for(int j = 0; j < byteSize; ++j) {
				value |= T.CreateTruncating(buff[i * byteSize + j]) << (j * 8);
			}
			builder.Add(value);
		}
		return builder.MoveToImmutable();
	}

	private async Task<Expr> ReadConstructor(string constructor, CancellationToken cancellationToken = default) {
		var argVisitor = new ConstructorArgumentVisitor(this, cancellationToken);

		while(!argVisitor.IsDone) {
			cancellationToken.ThrowIfCancellationRequested();

			await await ReadExprPlus(argVisitor, cancellationToken);
		}

		return new Expr.Constructor(constructor, argVisitor.Args.ToImmutable(), argVisitor.Kwargs.ToImmutable());
	}

	private sealed class ConstructorArgumentVisitor(ESExprBinaryReader reader, CancellationToken cancellationToken)
		: IExprPlusVisitor<ValueTask> {
		public ImmutableList<Expr>.Builder Args { get; } = ImmutableList.CreateBuilder<Expr>();
		public ImmutableDictionary<string, Expr>.Builder Kwargs { get; } = ImmutableDictionary.CreateBuilder<string, Expr>();
		public bool IsDone { get; private set; } = false;

		public async ValueTask VisitExpr(Expr expr) {
			Args.Add(expr);
		}

		public async ValueTask VisitConstructorEnd() {
			IsDone = true;
		}

		public async ValueTask VisitKeyword(BigInteger index) {
			var kw = reader.LookupStringTable(index);
			var value = await reader.Read(cancellationToken);
			Kwargs.Add(kw, value);
		}

		public async ValueTask VisitAppendedToStringTable() {
		}

		public async ValueTask VisitEndOfStream() {
			throw new EndOfStreamException();
		}
	}

	private async ValueTask<byte[]> ReadBytes(BigInteger tokenIntValue, CancellationToken cancellationToken = default) {
		if(tokenIntValue < 0 || tokenIntValue >= int.MaxValue) {
			throw new SyntaxException("Length too large");
		}

		int length = (int)tokenIntValue;
		// Assume nextByte is -1 here.
		byte[] buff = new byte[length];
		await stream.ReadExactlyAsync(buff, cancellationToken);
		return buff;
	}




	private async ValueTask<BinToken?> NextToken(CancellationToken cancellationToken = default) {
		int b = await Next(cancellationToken).ConfigureAwait(false);
		if(b < 0) {
			return null;
		}

		BinToken.TokenType? tokenType = (b & 0xE0) switch {
			0x00 => BinToken.TokenType.Constructor,
			0x20 => BinToken.TokenType.Int,
			0x40 => BinToken.TokenType.NegInt,
			0x60 => BinToken.TokenType.String,
			0x80 => BinToken.TokenType.StringPoolIndex,
			0xA0 => BinToken.TokenType.Array8,
			0xC0 => BinToken.TokenType.Keyword,
			_ => null,
		};

		BigInteger? intValue;
		if(tokenType is { } tt) {
			BigInteger i = b & 0x0F;
			if((b & 0x10) == 0x10) {
				i = await ReadInt(i, 4, cancellationToken).ConfigureAwait(false);
			}

			intValue = i;
		}
		else {
			tt = b switch {
				0xE0 => BinToken.TokenType.ConstructorEnd,
				0xE1 => BinToken.TokenType.True,
				0xE2 => BinToken.TokenType.False,
				0xE3 => BinToken.TokenType.Null0,
				0xE8 => BinToken.TokenType.Null1,
				0xE9 => BinToken.TokenType.Null2,
				0xEA => BinToken.TokenType.NullN,
				0xEC => BinToken.TokenType.Float16,
				0xE4 => BinToken.TokenType.Float32,
				0xE5 => BinToken.TokenType.Float64,
				0xED => BinToken.TokenType.Array16,
				0xEE => BinToken.TokenType.Array32,
				0xEF => BinToken.TokenType.Array64,
				0xF0 => BinToken.TokenType.Array128,
				0xE6 => BinToken.TokenType.ConstructorStartStringTable,
				0xE7 => BinToken.TokenType.ConstructorStartList,
				0xF1 => BinToken.TokenType.ConstructorStartMap,
				0xF2 => BinToken.TokenType.ConstructorStartSet,
				0xEB => BinToken.TokenType.AppendStringTable,
				_ => throw new SyntaxException(),
			};

			intValue = null;
		}

		return new BinToken(tt, intValue);
	}

	private async ValueTask<BigInteger> ReadInt(BigInteger acc, int bits, CancellationToken cancellationToken = default) {
		while(true) {
			int b = await Next(cancellationToken).ConfigureAwait(false);
			if(b < 0) {
				throw new SyntaxException();
			}

			acc |= (BigInteger)(b & 0x7F) << bits;
			bits += 7;

			if((b & 0x80) == 0) {
				return acc;
			}
		}
	}
}
